sympy/logic/boolalg.py

```diff
@@ -1637,6 +1807,188 @@ def POSform(variables, minterms, dontcares=None):
     return And(*[_convert_to_varsPOS(x, variables) for x in essential])
 
 
+def ANF(variables, truthvalues):
+    """
+    The ANF function converts the list of truth values to
+    Algebraic Normal Form (ANF).
+
+    The variables must be given as the first argument.
+
+    Return True, False, logical And function (i.e., the
+    "Zhegalkin monomial") or logical Xor function (i.e.,
+    the "Zhegalkin polynomial").
+
+    Formally a "Zhegalkin monomial" is the product (logical
+    And) of a finite set of distinct variables, including
+    the empty set whose product is denoted 1 (True).
+    A "Zhegalkin polynomial" is the sum (logical Xor) of a
+    set of Zhegalkin monomials, with the empty set denoted
+    by 0 (False).
+
+    Examples
+    ========
+    >>> from sympy.logic.boolalg import ANF
+    >>> ANF(['x'], [1, 0])
+    Xor(x, True)
+    >>> ANF(['x', 'y'], [0, 1, 1, 1])
+    Xor(x, y, And(x, y))
+
+    References
+    ==========
+
+    .. [2] en.wikipedia.org/wiki/Zhegalkin_polynomial
+
+    """
+    variables = [sympify(v) for v in variables]
+
+    coeffs = anf_coeffs(truthvalues)
+    terms = []
+
+    for i, t in enumerate(product([0, 1], repeat=len(variables))):
+        if coeffs[i] == 1:
+            terms.append(t)
+
+    return Xor(*[_convert_to_varsANF(x, variables) for x in terms],\
+            remove_true=False)
+
+
+def anf_coeffs(truthvalues):
+    """
+    Convert a list of truth values of some boolean expression
+    to the list of coefficients of the polynomial mod 2 (exclusive
+    disjunction) representing the boolean expression in ANF
+    (i.e., the "Zhegalkin polynomial").
+
+    There are 2^n possible Zhegalkin monomials in n variables, since
+    each monomial is fully specified by the presence or absence of
+    each variable.
+
+    A given monomial's presence or absence in a polynomial corresponds
+    to that monomial's coefficient being 1 or 0 respectively.
+
+    Examples
+    ========
+    >>> from sympy.logic.boolalg import anf_coeffs, monomial, Xor
+    >>> truthvalues = [0, 1, 1, 0, 0, 1, 0, 1]
+    >>> coeffs = anf_coeffs(truthvalues)
+    >>> coeffs
+    [0, 1, 1, 0, 0, 0, 1, 0]
+    >>> polynomial = Xor(*[monomial(k, ['a','b','c'])
+    ...         for k, coeff in enumerate(coeffs) if coeff==1])
+    >>> polynomial
+    Xor(b, c, And(a, b))
+
+    """
+
+    s = '{0:b}'.format(len(truthvalues))
+    n = len(s) - 1
+
+    if len(truthvalues) != 2**n:
+        raise ValueError('The number of truth values must be a power of two.')
+
+    coeffs = [[v] for v in truthvalues]
+
+    for i in range(n):
+        tmp = []
+        for j in range(2 ** (n-i-1)):
+            tmp.append(coeffs[2*j] +
+                list(map(lambda x, y: x^y, coeffs[2*j], coeffs[2*j+1])))
+        coeffs = tmp
+
+    return coeffs[0]
+
+
+def minterm(k, variables):
+    """
+    Return the k-th minterm.
```

Comment:
Function <code>minterm</code> returns a logical expression of _n_ variables that employs only the _complement_ operator and the _conjunction_ operator.

There are 2<sup>n</sup> minterms of n variables, since a variable in the minterm expression can be in either its direct or its complemented form—two choices per n variables.

Complementation pattern of the _k-th_ minterm can be deduced from the base-2 digits in the integer _k_.

Reference - http://en.wikipedia.org/wiki/Canonical_normal_form#Indexing_minterms

Function <code>minterm</code> is similar to function <code>BooleanMinterms</code> in Wolfram Mathematica http://reference.wolfram.com/language/ref/BooleanMinterms.html